app/domain/content/importers/metrics.rb

```diff
@@ -0,0 +1,117 @@
+module Content
+  module Importers
+    class Metrics
+      def initialize(content_ids)
+        @content_ids = content_ids
+        @start_date = 3.months.ago.to_date
+      end
+
+      def import
+        start_time = Time.now
+        metrics = @content_ids
+          .map { |content_id| fetch_metrics(content_id) }
+          .flatten
+        fetched_metrics_end = Time.now
+        puts "Fetched all in #{fetched_metrics_end - start_time}s"
+
+        puts "Writing metrics to database..."
+        # TODO: on_duplicate_key_update
+        Content::Metric.import(metrics)
+        puts "Imported all in #{Time.now - fetched_metrics_end}s"
+      end
+
+    private
+
+      def fetch_metrics(content_id)
+        puts "Fetching metrics for #{content_id} ..."
+        versions = fetch_versions(content_id)
+
+        metrics_by_version = {}
+
+        metrics = (@start_date..Date.today).map do |date|
+          version = versions.find { |v| Date.parse(v[:updated_at]) <= date }
+
+          if version.nil?
+            nil
+          else
+            metric_for_version = metrics_by_version.fetch(version[:user_facing_version]) do |version_number|
+              plain_text = Govspeak::Document.new(body(version)).to_text
+
+              metrics_by_version.store(
+                version_number,
+                title_length: version[:title].length,
+                reading_grade: Odyssey.smog(plain_text),
+              )
+            end
+
+            Content::Metric.new({
+              content_id: version[:content_id],
+              date: date,
+            }.merge(metric_for_version))
+          end
+        end
+
+        metrics.compact
+      end
+
+      def fetch_versions(content_id)
+        versions = []
+        preceding_version_number = nil
+
+        loop do
+          content_item = publishing_api.fetch(
+            content_id,
+            version: preceding_version_number,
+          )
+
+          versions << content_item
+
+          preceding_version_number = fetch_preceding_version_number(content_item)
+          updated_at_date = Date.parse(content_item[:updated_at]).to_date
+
+          break unless preceding_version_number && updated_at_date > @start_date
+        end
+
+        versions
+      end
+
+      def publishing_api
+        @publishing_api ||= Clients::PublishingAPI.new
+      end
+
+      def fetch_preceding_version_number(content_item)
+        content_item[:state_history]
+          .map { |version_number, _| Integer(version_number.to_s) }
+          .select { |version_number| version_number < content_item[:user_facing_version] }
+          .sort
+          .last
+      end
+
+      def body(content_item)
+        return "" unless content_item.dig(:details, :parts).present?
+        content_item[:details][:parts].reduce("") do |content_item_body, part|
+          content_item_body + if part[:body].present?
+                                part[:body].reduce("") do |part_body, body_fragment|
```

Comment:
What do you think about taking the second reduce function out into a seperate method to make this easier to read/understand?

`def concat_all_parts_of_content_item_body` or something like that?